### PR TITLE
Improve cuDNN (and XLA) deterministic convolution testing

### PR DESCRIPTION
**Pull request by duncanriach**
This PR improves the cuDNN (and XLA) deterministic convolution testing in the following ways:
* Because the existing tests evolved from TF1 API code but are now running on the TF2 eager API, they are operating such that they test nothing. They always pass, regardless of the state of the functionality that they intend to test. Specifically, code that would previously run a graph-op twice using the TF1 API and compare the results is instead running the op once using the TF2 eager API and then comparing the numerical result to itself, which always matches. This PR changes the test infrastructure to compare the result of a python function that is run twice in eager mode.
* The non-dilated convolution backprop tests are not exercising nondeterminism (when op-determinism is disabled) when using XLA JIT compilation. This PR adjusts the configurations such that nondeterminism is exercised (when op-determinism is disabled) both with and without XLA.
* This PR modifies some of the existing test cases to make them smaller, and therefore possibly makes them run faster, whilst robustly exercising nondeterminism (when op-determinism is disabled).
* This PR adds tests for `tf.nn.conv2d_transpose`.
* This PR adds a test that exercises forward nondeterminism (when op-determinism is disabled) for 3D convolution when using XLA JIT compilation.

### CODE REVIEW

**Comment by duncanriach** [comment]
cc @reedwm for review.

File: tensorflow/python/kernel_tests/nn_ops/cudnn_deterministic_base.py [hunk 1]
```diff
@@ -71,98 +73,117 @@ def _random_data_op(self, shape):
...
-        data_format='NCDHW',
-        dilations=dilations)
-    self._assert_reproducible(out_op)
+    self._assert_reproducible(lambda: nn_ops.conv3d(
+        in_op, filter_op, strides=[1, 1, 1, 1, 1], padding='VALID',
+        data_format='NCDHW', dilations=[1, 1, 2, 2, 2]))
 
+  # This test intends to exercise nondeterminism of convolution in the forward
+  # direction for XLA only. cuDNN does not have nondeterministic forward
+  # convolution algorithms.
```
Comment (reedwm) [inline tensorflow/python/kernel_tests/nn_ops/cudnn_deterministic_base.py:111]: This comment seems to imply that this test case tests that XLA convolutions are not deterministic even when determinism is enabled. I would phrase it something like "This test is primarily testing XLA, since cuDNN forward convolutions are always deterministic, even when determinism is not enabled. The convolution configuration tested is nondeterministic with XLA when determinism is disabled".

Also, does XLA not use cuDNN for convolutions? I am unfamiliar with the details of XLA codegen here.

File: tensorflow/python/kernel_tests/nn_ops/cudnn_deterministic_base.py [hunk 2]
```diff
@@ -71,98 +73,117 @@ def _random_data_op(self, shape):
...
+
+  # A configuration for this test could not be found that exercises
+  # nondeterminism on XLA.
+  @test_util.run_cuda_only
+  def testConvBackwardInputGradientWithDilations(self):
+    self.testConvBackwardInputGradient(rate=2)
+
+  # A 2D convolution configuration that exercises nondeterminism in the forward
+  # direction (i.e. tf.nn.conv2d) using XLA was not found and therefore no tests
+  # for the backprop of tf.nn.conv2d_transpose have been included.
```
Comment (reedwm) [inline tensorflow/python/kernel_tests/nn_ops/cudnn_deterministic_base.py:168]: For someone to understand this comment, they must know the backprop of conv2d_transpose is equivalent to the forward direction of conv2d. I would just drop this comment, or add the transpose backwards case. It's still useful to test determinism even if the op is unconditionally deterministic, since it's possible the implementation of the op later changes to be nondeterministic.

**Review by reedwm — CHANGES_REQUESTED** [review]
> Because the existing tests evolved from TF1 API code but are now running on the TF2 eager API, they are operating such that they test nothing

FYI, I think when the test is run internally, it uses the TF1 API and disables V2 behavior (or perhaps the CI is run twice, once with TF1 and once with TF2). But still good to make this change, for when the test is not run internally.

File: tensorflow/python/kernel_tests/nn_ops/cudnn_deterministic_base.py (hunk 1 above)
Comment (duncanriach) [reply tensorflow/python/kernel_tests/nn_ops/cudnn_deterministic_base.py:111]: I'll clarify the comment.

Regarding XLA's use of cuDNN for convolutions: historically, XLA definitely used cuDNN for convolutions, and I didn't think that had changed (based on a recent perusal of the relevant code). However, while playing around with these tests, I happened to find that this 3D convolution case exercises nondeterminism when op-determinism is not enabled, but only for the XLA auto-JIT case. I don't understand why it's like that.

File: tensorflow/python/kernel_tests/nn_ops/cudnn_deterministic_base.py (hunk 2 above)
Comment (duncanriach) [reply tensorflow/python/kernel_tests/nn_ops/cudnn_deterministic_base.py]: I have removed the comment and added tests for reproducibility of the gradients (wrt both `filters` and `input`) of `tf.nn.conv2d_transpose`. Interestingly, it seems that the backprop to `filters` for `tf.nn.conv2d_transpose` is nondeterministic when determinism is not enabled, presumably because it uses the same functionality as backprop to `filters` for `tf.nn.conv2d`, due to the symmetry.

**Comment by duncanriach** [comment]
> > Because the existing tests evolved from TF1 API code but are now running on the TF2 eager API, they are operating such that they test nothing
> 
> FYI, I think when the test is run internally, it uses the TF1 API and disables V2 behavior (or perhaps the CI is run twice, once with TF1 and once with TF2). But still good to make this change, for when the test is not run internally.

As for many other tests, I don't believe that these tests are compatible (any longer) with the TF1 API. Recall that [`run_in_graph_and_eager_modes`](https://github.com/tensorflow/tensorflow/blob/e3ab82b205e72716dd00baa551a611d606ec31cc/tensorflow/python/framework/test_util.py#L1422-L1526) can be used to specify that a test should (and therefore can) be run on both APIs. [`run_deprecated_v1` == `deprecated_graph_mode_only`](https://github.com/tensorflow/tensorflow/blob/e3ab82b205e72716dd00baa551a611d606ec31cc/tensorflow/python/framework/test_util.py#L1588-L1637) was used to prevent attempts to run old tests written for the TF1 API on the TF2/eager API.

File: tensorflow/python/kernel_tests/nn_ops/cudnn_deterministic_base.py [hunk 3]
```diff
@@ -183,7 +182,71 @@ def testConvTransposeForward(self, rate=1):
...
+            data_format='NHWC', dilations=[1, rate, rate, 1])
+        gradient_injector_output = op_output * upstream_gradients
+      return tape.gradient(gradient_injector_output, [filter_op])[0]
+
+    self._assert_reproducible(gradient)
+
+  # A configuration for this test could not be found that exercises
+  # nondeterminism when using XLA with determinism not enabled.
+  @test_util.run_cuda_only
+  def testConvTransposeBackwardFilterGradientWithDilations(self, rate=1):
```
Comment (reedwm) [inline tensorflow/python/kernel_tests/nn_ops/cudnn_deterministic_base.py:218]: You should remove the `rate` argument here, since you pass it below. And same for `testConvTransposeBackwardInputGradientWithDilations`

**Review by reedwm — CHANGES_REQUESTED** [review]

File: tensorflow/python/kernel_tests/nn_ops/cudnn_deterministic_base.py (hunk 3 above)
Comment (duncanriach) [reply tensorflow/python/kernel_tests/nn_ops/cudnn_deterministic_base.py]: Oops. I copy-pasted without thinking enough. Thanks for catching that.

**Review by reedwm — APPROVED** [review]
